crates/primitives/src/transaction/mod.rs

```diff
@@ -582,9 +676,28 @@ impl TransactionSigned {
         &self.signature
     }
 
-    /// Transaction hash. Used to identify transaction.
+    /// Transaction hash. Used to identify transaction. Calculates it if it's not present, but does
+    /// not store it.
     pub fn hash(&self) -> TxHash {
-        self.hash
+        if let Some(hash) = self.hash {
+            return hash
+        }
+        self.recalculate_hash()
+    }
+
+    /// Transaction hash. Used to identify transaction. Calculates it if it's not present and stores
+    /// it for future use.
+    pub fn hash_mut(&mut self) -> TxHash {
+        if let Some(hash) = self.hash {
+            return hash
+        }
+        self.hash = Some(self.recalculate_hash());
+        self.hash.expect("qed")
```

Comment:
```rust
self.hash.get_or_insert_with(|| self.recalculate_hash());
```
this cannot work though, since `self.hash` is a mutable borrow, which is followed by a disallowed immutable one `|| self.recalculate_hash`